### PR TITLE
adds support for gpg commit signing (fixes #1018)

### LINKED ISSUE #1018: Support gpg commit signing
nodegit does not appear to have support yet for gpg commit signing. Perhaps I have missed it in the documentation, but there does not appear to be support for this quite yet. Do you have plans to support it some time soon? If not, would you be willing to specify the API you would like to see so others could contribute? 

Thanks.

### PR DESCRIPTION
**Pull request by dabutvin**
This wraps the libgit commit signing methods.

~~I couldn't get to `git_commit_extract_signature` as it's not available until 0.24
https://libgit2.github.com/libgit2/#HEAD/group/commit/git_commit_extract_signature~~

### CODE REVIEW

**Comment by rcjsuen** [comment]
How does this pull request add signing support? It seems like it's just adding a new convenience `Repository.createCommitWithSignature` function that wraps around the existing `Commit.createWithSignature`.

**Comment by dabutvin** [comment]
That's exactly right - it wraps the Commit.createWithSignature method.
I thought that's what #1018 was calling for - sorry for the mixup.

Would you rather see this method extend `Repository.createCommit` to take the private key and password additional parameters and do the signing right here in this library?

```
Repository.prototype.createCommitWithSignature = function(
    updateRef, author, committer, message, tree, parents, privateKey, password, callback)
```

Or do you have something else in mind?

**Comment by rcjsuen** [comment]
@dabutvin I think that's what people want but you should ask in #1018 to clarify and see what their desired API would look like.

**Comment by implausible** [comment]
Make sure to enable support for `git_commit_extract_signature` for this PR by updating https://github.com/nodegit/nodegit/blob/master/generate/input/descriptor.json#L525

**Comment by dabutvin** [comment]
@implausible awesome! I'll give that a try

**Comment by dabutvin** [comment]
@implausible I'm having some trouble calling extractSignature.
It accepts some `GitBuf`s and I can't seem to instantiate them.

Do you know a way to handle these buffers to write the signature and signed_data to?

Here's the generated .cc
```
/*
   * @param Buffer signature
   * @param Buffer signed_data
   * @param Repository repo
   * @param Oid commit_id
   * @param String field
 */
NAN_METHOD(GitCommit::ExtractSignature) {

  if (info.Length() == 0 || !info[0]->IsObject()) {
    return Nan::ThrowError("Buffer signature is required.");
  }

  if (info.Length() == 1 || !info[1]->IsObject()) {
    return Nan::ThrowError("Buffer signed_data is required.");
  }

  if (info.Length() == 2 || !info[2]->IsObject()) {
    return Nan::ThrowError("Repository repo is required.");
  }

  if (info.Length() == 3
    || (!info[3]->IsObject() && !info[3]->IsString())) {
    return Nan::ThrowError("Oid commit_id is required.");
  }
  if (info.Length() == 4 || !info[4]->IsString()) {
    return Nan::ThrowError("String field is required.");
  }

  if (info.Length() == 5 || !info[5]->IsFunction()) {
    return Nan::ThrowError("Callback is required and must be a Function.");
  }

  ExtractSignatureBaton* baton = new ExtractSignatureBaton;

  baton->error_code = GIT_OK;
  baton->error = NULL;

// start convert_from_v8 block
  git_buf * from_signature = NULL;

  from_signature = GitBufConverter::Convert(info[0]);
// end convert_from_v8 block
  baton->signature = from_signature;
// start convert_from_v8 block
  git_buf * from_signed_data = NULL;

  from_signed_data = GitBufConverter::Convert(info[1]);
// end convert_from_v8 block
  baton->signed_data = from_signed_data;
// start convert_from_v8 block
  git_repository * from_repo = NULL;
from_repo = Nan::ObjectWrap::Unwrap<GitRepository>(info[2]->ToObject())->GetValue();
// end convert_from_v8 block
  baton->repo = from_repo;
// start convert_from_v8 block
  git_oid * from_commit_id = NULL;
  if (info[3]->IsString()) {
    // Try and parse in a string to a git_oid
    String::Utf8Value oidString(info[3]->ToString());
    git_oid *oidOut = (git_oid *)malloc(sizeof(git_oid));

    if (git_oid_fromstr(oidOut, (const char *) strdup(*oidString)) != GIT_OK) {
      free(oidOut);

      if (giterr_last()) {
        return Nan::ThrowError(giterr_last()->message);
      } else {
        return Nan::ThrowError("Unknown Error");
      }
    }

    from_commit_id = oidOut;
  }
  else {
from_commit_id = Nan::ObjectWrap::Unwrap<GitOid>(info[3]->ToObject())->GetValue();
  }
// end convert_from_v8 block
  baton->commit_id = from_commit_id;
  baton->commit_idNeedsFree = info[3]->IsString();
// start convert_from_v8 block
  const char * from_field = NULL;

  String::Utf8Value field(info[4]->ToString());
  // malloc with one extra byte so we can add the terminating null character C-strings expect:
  from_field = (const char *) malloc(field.length() + 1);
  // copy the characters from the nodejs string into our C-string (used instead of strdup or strcpy because nulls in
  // the middle of strings are valid coming from nodejs):
  memcpy((void *)from_field, *field, field.length());
  // ensure the final byte of our new string is null, extra casts added to ensure compatibility with various C types
  // used in the nodejs binding generation:
  memset((void *)(((char *)from_field) + field.length()), 0, 1);
// end convert_from_v8 block
  baton->field = from_field;

  Nan::Callback *callback = new Nan::Callback(v8::Local<Function>::Cast(info[5]));
  ExtractSignatureWorker *worker = new ExtractSignatureWorker(baton, callback);
  if (!info[0]->IsUndefined() && !info[0]->IsNull())
    worker->SaveToPersistent("signature", info[0]->ToObject());
  if (!info[1]->IsUndefined() && !info[1]->IsNull())
    worker->SaveToPersistent("signed_data", info[1]->ToObject());
  if (!info[2]->IsUndefined() && !info[2]->IsNull())
    worker->SaveToPersistent("repo", info[2]->ToObject());
  if (!info[3]->IsUndefined() && !info[3]->IsNull())
    worker->SaveToPersistent("commit_id", info[3]->ToObject());
  if (!info[4]->IsUndefined() && !info[4]->IsNull())
    worker->SaveToPersistent("field", info[4]->ToObject());

  AsyncLibgit2QueueWorker(worker);
  return;
}
```

**Comment by cjhoward92** [comment]
You can circumvent the `GitBuf` with a `new String()`. [See here for an example](https://github.com/nodegit/nodegit/blob/master/test/tests/filter.js#L1035)

**Comment by cjhoward92** [comment]
Let me know if you need any other help on this. I would really like to get GPG signing in sooner rather than later.

**Comment by dabutvin** [comment]
@cjhoward92 thanks. I did give that a try, but I need to read the value back out.

```
var buffer1 = new String();
var buffer2 = new String();

var responseCode = extractSignature(buffer1, buffer2, repo, oid, 'gpgsig');
```

I can get `responseCode` to go to 0 (success) with the right parameters, but `buffer1` is never populated.

I'm not sure what I am missing.

File: test/tests/commit.js [hunk 1]
```diff
@@ -199,6 +199,46 @@ describe("Commit", function() {
...
+      "a simple commit which works\n";
+
+    return NodeGit.Repository.open(reposPath)
+    .then(function(repo) {
+      return repo.createCommitWithSignature(
+        commit_content,
+        signature);
+    })
+    .then(function(commitId) {
+      assert.equal(expectedCommitId, commitId);
```
Comment (cjhoward92) [inline test/tests/commit.js]: I know you are still working, but I would like to see a test that verifies the signature in the `gpgsig` field of the commit. That will definitely lend credence to fact that we have GPG signing.

**Comment by cjhoward92** [comment]
Oh, I see. I didn't realize you were talking about the `ExtractSignature` function. Yeah, strings won't work. Does a `Buffer` object from node work? We may need to add some code to allocate a buffer for us. Let me look around and see what we have that does that already, if anything.

I think the final signature will probably look different since we need to return some of this stuff...

**Comment by dabutvin** [comment]
yes it makes sense. I'll try out generating with the buffer and see if I can get it to work.

For what to return.
There are 2 outputs of ExtractSignature  `signature` and `signed_data`

We could return an object that has both, or just return the signature.

`var result = commit.extractSignature()`

and `result` could be 

```
{
    'signature': 'asiudqw....',
    'signed_data': 'tree asdads....'
}
```

or `result` could be just the signature

```
asiudqw...
```

the method I was planning to add to lib/commit.js would look something like this, but with the promises

```
Commit.prototype.extractSignature = function() {
    var signatureBuffer = new String();
    var signedDataBuffer = new String();

    NodeGit.Commit.extractSignature(signatureBuffer, signedDataBuffer, this.repo, this.id(), "gpgsig");

    return signatureBuffer.toString();
}
```

**Comment by cjhoward92** [comment]
Do you know what the difference between `signature` and `signature_data` is? What would we need `signature_data` for? I haven't used GPG enough or looked at libgit2 in much detail, so I am not sure what we will be using `ExtractSignature` for. I assume signature verification, but I don't really know.

Overall I like your ideas! We just need to find the best path for you to be able to execute. For now, you can try setting just the `signature` as the return for that function and see if you get it working. If so, then we can move to the next step of determining the best way to return all of the data.

**Comment by dabutvin** [comment]
sounds good. I'll try to get something working.

I believe the intention of ExtractSignature is so that someone could take the signed_data and the signature and verify it against a public key

signature is the actual signature produced from signing the commit.
signed_data is the content that was signed 

This is the full commit:

```
"tree 6b79e22d69bf46e289df0345a14ca059dfc9bdf6\n\
parent 34734e478d6cf50c27c9d69026d93974d052c454\n\
author Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
committer Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
gpgsig -----BEGIN PGP SIGNATURE-----\n\
 Version: GnuPG v1.4.12 (Darwin)\n\
 \n\
 iQIcBAABAgAGBQJQ+FMIAAoJEH+LfPdZDSs1e3EQAJMjhqjWF+WkGLHju7pTw2al\n\
 o6IoMAhv0Z/LHlWhzBd9e7JeCnanRt12bAU7yvYp9+Z+z+dbwqLwDoFp8LVuigl8\n\
 JGLcnwiUW3rSvhjdCp9irdb4+bhKUnKUzSdsR2CK4/hC0N2i/HOvMYX+BRsvqweq\n\
 AsAkA6dAWh+gAfedrBUkCTGhlNYoetjdakWqlGL1TiKAefEZrtA1TpPkGn92vbLq\n\
 SphFRUY9hVn1ZBWrT3hEpvAIcZag3rTOiRVT1X1flj8B2vGCEr3RrcwOIZikpdaW\n\
 who/X3xh/DGbI2RbuxmmJpxxP/8dsVchRJJzBwG+yhwU/iN3MlV2c5D69tls/Dok\n\
 6VbyU4lm/ae0y3yR83D9dUlkycOnmmlBAHKIZ9qUts9X7mWJf0+yy2QxJVpjaTGG\n\
 cmnQKKPeNIhGJk2ENnnnzjEve7L7YJQF6itbx5VCOcsGh3Ocb3YR7DMdWjt7f8pu\n\
 c6j+q1rP7EpE2afUN/geSlp5i3x8aXZPDj67jImbVCE/Q1X9voCtyzGJH7MXR0N9\n\
 ZpRF8yzveRfMH8bwAJjSOGAFF5XkcR/RNY95o+J+QcgBLdX48h+ZdNmUf6jqlu3J\n\
 7KmTXXQcOVpN6dD3CmRFsbjq+x6RHwa8u1iGn+oIkX908r97ckfB/kHKH7ZdXIJc\n\
 cpxtDQQMGYFpXK/71stq\n\
 =ozeK\n\
 -----END PGP SIGNATURE-----\n\
\n\
a simple commit which works\n"
```

this is the signature:

```
-----BEGIN PGP SIGNATURE-----\n\
" Version: GnuPG v1.4.12 (Darwin)\n\
 \n\
 iQIcBAABAgAGBQJQ+FMIAAoJEH+LfPdZDSs1e3EQAJMjhqjWF+WkGLHju7pTw2al\n\
 o6IoMAhv0Z/LHlWhzBd9e7JeCnanRt12bAU7yvYp9+Z+z+dbwqLwDoFp8LVuigl8\n\
 JGLcnwiUW3rSvhjdCp9irdb4+bhKUnKUzSdsR2CK4/hC0N2i/HOvMYX+BRsvqweq\n\
 AsAkA6dAWh+gAfedrBUkCTGhlNYoetjdakWqlGL1TiKAefEZrtA1TpPkGn92vbLq\n\
 SphFRUY9hVn1ZBWrT3hEpvAIcZag3rTOiRVT1X1flj8B2vGCEr3RrcwOIZikpdaW\n\
 who/X3xh/DGbI2RbuxmmJpxxP/8dsVchRJJzBwG+yhwU/iN3MlV2c5D69tls/Dok\n\
 6VbyU4lm/ae0y3yR83D9dUlkycOnmmlBAHKIZ9qUts9X7mWJf0+yy2QxJVpjaTGG\n\
 cmnQKKPeNIhGJk2ENnnnzjEve7L7YJQF6itbx5VCOcsGh3Ocb3YR7DMdWjt7f8pu\n\
 c6j+q1rP7EpE2afUN/geSlp5i3x8aXZPDj67jImbVCE/Q1X9voCtyzGJH7MXR0N9\n\
 ZpRF8yzveRfMH8bwAJjSOGAFF5XkcR/RNY95o+J+QcgBLdX48h+ZdNmUf6jqlu3J\n\
 7KmTXXQcOVpN6dD3CmRFsbjq+x6RHwa8u1iGn+oIkX908r97ckfB/kHKH7ZdXIJc\n\
 cpxtDQQMGYFpXK/71stq\n\
 =ozeK\n\
 -----END PGP SIGNATURE-----"
```

This is signed_data:

```
"tree 6b79e22d69bf46e289df0345a14ca059dfc9bdf6\n\
parent 34734e478d6cf50c27c9d69026d93974d052c454\n\
author Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
committer Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
\n\
a simple commit which works\n";
```

**Comment by cjhoward92** [comment]
Cool, that makes sense. Yeah, we should probably get that in. Let me know if you want any extra help or just a second set of eyes!

Thanks again for working on this. I know a lot of people want GPG in NodeGit so you will be a hero!

**Comment by implausible** [comment]
@dabutvin I am opening up a PR shortly that will get git_commit_extract_signature built into nodegit for you. You'll want to rebase on that. I will post here when it's open.

**Comment by implausible** [comment]
@dabutvin https://github.com/nodegit/nodegit/pull/1458 this PR might help you out :).

**Comment by dabutvin** [comment]
@implausible I updated to verify the signature from the signed commit using the new `commit.getSignature` method.

**Comment by dabutvin** [comment]
Can you work your magic in descriptor.json for `git_commit_create_buffer`?

This is what I am trying to set up in repository.js

onSignature would be a function that is up to the caller to implement

```
Repository.prototype.createCommitWithSignature = function(
    author, committer, message, tree, parents, signature_field, onSignature, callback) {

  var repo = this;
  var promises = [];

  parents = parents || [];

  promises.push(repo.getTree(tree));

  parents.forEach(function(parent) {
    promises.push(repo.getCommit(parent));
  });

  return Promise.all(promises).then(function(results) {
    tree = results[0];

    // Get the normalized values for our input into the function
    var parentsLength = parents.length;
    parents = [];

    for (var i = 0; i < parentsLength; i++) {
      parents.push(results[i + 1]);
    }

    return Commit.createBuffer(
      repo,
      author,
      committer,
      null /* use default message encoding */,
      message,
      tree,
      parents.length,
      parents
    );
  }).then(function(commit_content) {
    return Commit.createWithSignature(
        repo,
        commit_content,
        onSignature(commit_content),
        "gpgsig");
  }, callback);
};
```

File: test/tests/commit.js [hunk 2]
```diff
@@ -720,7 +719,67 @@ describe("Commit", function() {
   });
 
   describe("Commit's Signature", function() {
-    it.only("Can retrieve the gpg signature from a commit", function() {
+
+    it("Can create a signed commit", function() {
+      var expectedCommitId = "b78570159824e6fc161418a626a646933a10ea89";
+
+      // https://github.com/libgit2/libgit2/blob/master/tests/commit/write.c#L362-L384
```
[the remark this answers is not in the document]
Comment (rcjsuen) [reply test/tests/commit.js]: Perhaps you can replace it with these Axosoft signatures that were merged in from #1458?

https://github.com/nodegit/nodegit/blob/210d02a47e080f4872cc17665ac08feb2e204680/test/tests/commit.js#L723-L740

**Comment by implausible** [comment]
One thing to keep in mind with the onSignature callback is that it may be asynchronous. I would change the flow in your code to account for that.

**Comment by implausible** [comment]
@dabutvin https://github.com/nodegit/nodegit/pull/1481

**Comment by dabutvin** [comment]
@implausible thanks for exposing the `createCommitBuffer` 
I updated the branch

**Comment by chapterjason** [comment]
No changes for a while? Still working on it? @dabutvin

**Comment by dabutvin** [comment]
I forgot all about this actually. I'll rebase the commit now - it looks finished, but there is a broken build.

**Comment by dabutvin** [comment]
it passed!

File: lib/repository.js [hunk 3]
```diff
@@ -622,6 +622,74 @@ Repository.prototype.createCommitBuffer = function(
...
+ */
+Repository.prototype.createCommitWithSignature = function(
+    author,
+    committer,
+    message,
+    tree,
+    parents,
+    signature_field,
+    onSignature,
+    callback) {
```
Comment (implausible) [inline lib/repository.js]: I don't think we need a callback param here because we're using a promise already, can we remove this param?

Comment (dabutvin) [reply lib/repository.js]: :) I'm happy to remove it if it's not needed.

It's referenced on line 690. Is it not needed for the rejection call?
I see other places in this file with a similar pattern. like `Repository.getReferences`

Comment (implausible) [reply lib/repository.js]: I think those are old patterns before the library was heavily promisified. It's more of a remnant of old code style and a smell in new code.

Yes, please remove the ref to callback from 690 and let's get this merged!

Comment (dabutvin) [reply lib/repository.js]: sounds good. I pushed a commit removing the callback

**Comment by dabutvin** [comment]
@implausible not sure what happened to the build, but the documentation config looks like it hung.
Maybe it just needs to be restarted?

**Comment by dabutvin** [comment]
@implausible check out newline I appended to the end of the commit buffer and let me know what you think.

I was doing some end to end testing of the signing and verification and found that when verifying the signature from the commit content, this newline was required to get an exact match with the commit content.

**Comment by implausible** [comment]
Thanks for the reminder!

**Comment by implausible** [comment]
This is great. We're also working on adding: https://github.com/libgit2/libgit2/pull/4913. Hopefully the next release of nodegit will include signing for rebase and commit.

**Pull request merged**